web-common/src/features/sources/modal/submitAddDataForm.ts

```diff
@@ -141,14 +111,90 @@ export async function submitAddDataForm(
     createOnly: false,
   });
 
-  // Update the `.env` file
+  // Check if .env file exists before we update it
+  let envFileExisted = false;
+  try {
+    const envFile = await queryClient.fetchQuery({
+      queryKey: getRuntimeServiceGetFileQueryKey(instanceId, { path: ".env" }),
+      queryFn: () => runtimeServiceGetFile(instanceId, { path: ".env" }),
+    });
+    envFileExisted = !!envFile.blob;
+  } catch (error) {
+    // If file doesn't exist, envFileExisted remains false
+    if (!error?.response?.data?.message?.includes("no such file")) {
+      throw error; // Re-throw if it's a different error
+    }
+  }
+
+  // Create or update the `.env` file
+  const newEnvBlob = await updateDotEnvWithSecrets(
+    queryClient,
+    connector,
+    formValues,
+    "connector",
+  );
   await runtimeServicePutFile(instanceId, {
     path: ".env",
-    blob: await updateDotEnvWithSecrets(queryClient, connector, formValues),
+    blob: newEnvBlob,
     create: true,
     createOnly: false,
   });
 
+  /**
+   * Test the connection to validate the connector configuration
+   */
+  const result = await testOLAPConnector(
+    instanceId,
+    newConnectorFilePath,
+    newConnectorName,
+  );
+
+  /**
+   * Rollback all changes if the connection test fails
+   */
+  if (!result.success) {
+    // Clean-up the `connector.yaml` file
+    await runtimeServiceDeleteFile(instanceId, {
+      path: newConnectorFilePath,
+    });
+
+    // Clean-up the `.env` file
+    if (!envFileExisted) {
```

Comment:
How long could the test take? Do we need to worry about overriding unintended `.env` changes if it takes a long time?

Reply:
It could take 20-30s until the backend context deadline is exceeded. When this is happening, I've prevented the dialog from closing like so:
```
<Dialog.Root
    open
    onOpenChange={async (open) => {
      if (!open) {
        await onCancelDialog();
      }
    }}
    closeOnEscape={!isSubmittingForm}
    closeOnOutsideClick={!isSubmittingForm}
  >
```